apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,373 @@
+  _relayMessageEvent: function bt__relayMessageEvent(msg, transferInfo) {
+    this._setProfileConnected(this.Profiles.OPP, true);
+    var evt = document.createEvent('CustomEvent');
+    evt.initCustomEvent(msg,
+      /* canBubble */ true, /* cancelable */ false,
+      {transferInfo: transferInfo});
+    window.dispatchEvent(evt);
+  },
+
+  /**
+   * Maintain connect stat of supported profiles.
+   *
+   * @private
+   * @param  {Object} adapter bluetooth adapter
+   */
+  _updateProfileStat: function bt_updateProfileStat(adapter) {
+    /* for v1, we only support two use cases for bluetooth connection:
+     *   1. connecting with a headset
+     *   2. transfering a file to/from another device
+     * So we need to listen to corresponding events to know we are (aren't)
+     * connected, then summarize to an event and dispatch to StatusBar
+     */
+    // In headset connected case:
+    adapter.addEventListener('hfpstatuschanged', (evt) => {
+      this._setProfileConnected(this.Profiles.HFP, evt.status);
+    });
+
+    adapter.addEventListener('a2dpstatuschanged', (evt) => {
+      this._setProfileConnected(this.Profiles.A2DP, evt.status);
+    });
+
+    adapter.addEventListener('scostatuschanged', (evt) => {
+      this._setProfileConnected(this.Profiles.SCO, evt.status);
+    });
+  },
+
+  /**
+   * BT APIv2: Watch 'onattributechanged' event from
+   * mozBluetooth.defaultAdapter for updating state information.
+   *
+   * @private
+   * @param  {Object} evt event object
+   */
+  _bluetoothAdapterAttrChangeHandler:
+    function bt__bluetoothAdapterAttrChangeHandler(evt) {
+      for (var i in evt.attrs) {
+        switch (evt.attrs[i]) {
+          case 'state':
+            if (this._adapter.state === 'enabled') {
+              this._isEnabled = true;
+              window.dispatchEvent(new CustomEvent('bluetooth-enabled'));
+              // this.icon && this.icon.update();
+            } else if (this._adapter.state === 'disabled') {
+              this._isEnabled = false;
+              window.dispatchEvent(new CustomEvent('bluetooth-disabled'));
+              // this.icon && this.icon.update();
+            }
+            break;
+          default:
+            break;
+        }
+      }
+  },
+
+  /**
+   * Watch 'onattributechanged' event from mozBluetooth for updating default
+   * adapter information.
+   *
+   * 'onattributechanged' event description:
+   * A handler to trigger when bluetooth manager's only property
+   * defaultAdapter has changed.
+   *
+   * @private
+   * @param  {Object} evt event object
+   */
+  _bluetoothAttrChangeHandler: function bt__bluetoothAttrChangeHandler(evt) {
+    for (var i in evt.attrs) {
+      switch (evt.attrs[i]) {
+        case 'defaultAdapter':
+          // Default adapter attribute change.
+          // Usually, it means that we reach new default adapter.
+          this._adapter = this._bluetooth.defaultAdapter;
+          this._adapter.addEventListener('attributechanged',
```

Comment:
in new implementation the defaultAdapter listener would be call once and being cached. Once the adapter is removed we will set adapter to `null`, so all event listener should be recycled